packages/jest-runner/src/run_test.js

```diff
@@ -95,39 +95,35 @@ export default function runTest(
     mapCoverage: globalConfig.mapCoverage,
   });
   const start = Date.now();
-  return testFramework(globalConfig, config, environment, runtime, path)
-    .then((result: TestResult) => {
-      const testCount =
-        result.numPassingTests +
-        result.numFailingTests +
-        result.numPendingTests;
-      result.perfStats = {end: Date.now(), start};
-      result.testFilePath = path;
-      result.coverage = runtime.getAllCoverageInfo();
-      result.sourceMaps = runtime.getSourceMapInfo();
-      result.console = testConsole.getBuffer();
-      result.skipped = testCount === result.numPendingTests;
-      result.displayName = config.displayName;
-      return result;
-    })
-    .then(
-      result =>
-        Promise.resolve().then(() => {
-          environment.dispose();
-          if (globalConfig.logHeapUsage) {
-            if (global.gc) {
-              global.gc();
-            }
-            result.memoryUsage = process.memoryUsage().heapUsed;
-          }
-
-          // Delay the resolution to allow log messages to be output.
-          return new Promise(resolve => setImmediate(() => resolve(result)));
-        }),
-      err =>
-        Promise.resolve().then(() => {
-          environment.dispose();
-          throw err;
-        }),
+  await environment.setup();
+  try {
+    const result: TestResult = await testFramework(
+      globalConfig,
+      config,
+      environment,
+      runtime,
+      path,
     );
+    const testCount =
+      result.numPassingTests + result.numFailingTests + result.numPendingTests;
+    result.perfStats = {end: Date.now(), start};
+    result.testFilePath = path;
+    result.coverage = runtime.getAllCoverageInfo();
+    result.sourceMaps = runtime.getSourceMapInfo();
+    result.console = testConsole.getBuffer();
+    result.skipped = testCount === result.numPendingTests;
+    result.displayName = config.displayName;
+    await environment.teardown();
```

Comment:
I think this diff needs a little bit more work. With the current code, if `teardown` throws, we'll throw here and then call teardown again in the catch block. That's not the same as what we used to do before.

Reply:
will this work?
```javascript
  try {

    snip...

    try {
      await environment.teardown();
    } catch (err) {
      throw Object.assign({}, err, {name: 'TeardownError'});
    }
    if (globalConfig.logHeapUsage) {
      if (global.gc) {
        global.gc();
      }
      result.memoryUsage = process.memoryUsage().heapUsed;
    }
    // Delay the resolution to allow log messages to be output.
    return new Promise(resolve => setImmediate(() => resolve(result)));
  } catch (err) {
    if (err.name !== 'TeardownError') {
      await environment.teardown();
    }
    throw err;
  }
```

Reply:
What about using finally?

Reply:
I thought you wanna dispose something before collecting the memoryUsage 🤔

Reply:
finally runs before returning from the function, so it should be good.

Reply:
like this?
```javascript
try {
    const result: TestResult = await testFramework(
      globalConfig,
      config,
      environment,
      runtime,
      path,
    );
    const testCount =
      result.numPassingTests + result.numFailingTests + result.numPendingTests;
    result.perfStats = {end: Date.now(), start};
    result.testFilePath = path;
    result.coverage = runtime.getAllCoverageInfo();
    result.sourceMaps = runtime.getSourceMapInfo();
    result.console = testConsole.getBuffer();
    result.skipped = testCount === result.numPendingTests;
    result.displayName = config.displayName;
    if (globalConfig.logHeapUsage) {
      if (global.gc) {
        global.gc();
      }
      result.memoryUsage = process.memoryUsage().heapUsed;
    }
    // Delay the resolution to allow log messages to be output.
    return new Promise(resolve => setImmediate(() => resolve(result)));
  } catch (err) {
    throw err;
  } finally {
    await environment.teardown();
  }
```

Reply:
I think that works for me, yes.